Add: sysProfileFocusChangeEvent

SlySven opened the pull request:
~~This is~~ Originally an attempt to close #5960 by detecting when the currently active host changes when multi-playing and issuing an event to replace the ones lost when I revised another part of the codebase to NOT issue a NAWS report or a `sysWindowResizeEvent` when the size of the main window was NOT actually changed but just shown/hidden. Subsequent other work revealed that that issue had a different cause. Instead this PR was expanded to help with #6273, #6449 and #6450 and generally clean up on matters relating to switching between profiles.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Add: sysProfileFocusChangeEvent

This is an attempt to close #5960 by detecting when the currently active
host changes when multi-playing and issuing an event to replace the ones
lost when I revised another part of the codebase to NOT issue a NAWS report
or a `sysWindowResizeEvent` when the size of the main window was NOT
actaully changed but just shown/hidden.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
This doesn't work yet - my attempt at getting the Geyser system to repaint the windows does not work to cure the problem that the sample code by the OP in the associated issue demonstrates. It needs someone with more Lua/Geyser-fu to work on it - so I think I need to call :mega:  in @demonnic ...!

Kebap commented:
Related discussion in #6273 suggest maybe a combined first step would be to refactor/bring to a united position all known ways to change the current profile. Afterwards, title can be updated accordingly (my issue) and the event may be issued (your issue).

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven committed:
Revise: put event using code inside a zero time timer

This seems to be needed to get around some odd behaviour when showing/
hiding gauges because something is not totally correct with the reported
size of the main window.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven committed:
Revert: undo changes to GeyserReposition.lua

The lua code put in there was trying to use this new event to force a
redraw of Geyser widgets when a profile hidden during multi-playing when
multi-view was not in use. This was an attempt to solve #5960. However it
seems the problem was with faulty reading of the `TMainConsole` size whilst
it was being shown. That appears to have been solved by a different PR:
#6438 that is waiting to be merged into the code-base. As such the tweaking
with the lua file will not be required.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven commented:
Looking into how Mudlet is currently operating I realise that @Kebap has a very valid point - and I need to hook up all the uses of:
* `(void) mudlet::activateProfile(Host*)`
* `(void) mudlet::mpTabBar::setCurrentIndex(int)`

as well as the existing `(void) mudlet::slot_tabChanged(int)` code I put together. to generate the new `sysProfileChangeEvent`, ~~converting this PR back to a **draft** until I have~~ *it is now* got this sorted out...

SlySven committed:
Fix: put focus in right place after clicking in places

This should close #6272 by better handling of the various
events raised by different widgets when they are clicked
on with the mouse.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven commented on src/mudlet.cpp:
```diff
@@ -671,9 +672,6 @@ mudlet::mudlet()
     connect(this, &mudlet::signal_menuBarVisibilityChanged, this, &mudlet::slot_updateShortcuts);
     connect(this, &mudlet::signal_hostCreated, this, &mudlet::slot_assignShortcutsFromProfile);
     connect(this, &mudlet::signal_profileActivated, this, &mudlet::slot_assignShortcutsFromProfile);
-    connect(this, &mudlet::signal_tabChanged, this, [=]() {
```
This connection is no longer needed  because it is duplicated by the one on the previous line.

SlySven commented on src/mudlet.h:
```diff
@@ -514,7 +515,6 @@ public slots:
 
 
 private slots:
-    void slot_tabChanged(int);
```
Can't be private any more as it is now called from other methods.

vadi2 commented:
How will this work?

Would any code that reacted to `sysWindowResizeEvent` now needs to react to `sysProfileFocusChangeEvent` as well or things won't look right?

Kebap commented:
> This is an attempt to close https://github.com/Mudlet/Mudlet/issues/5960

That issue was already solved. As it turned out, the window size was calculated wrong.

Which issue does this PR fix now?

Kebap commented:
#6273 seems improved / solved with this PR.
#6449 seems unaffected / not solved with this.

Please let me know what else is there to test.

SlySven commented:
> How will this work?
> 
> Would any code that reacted to `sysWindowResizeEvent` now needs to react to `sysProfileFocusChangeEvent` as well or things won't look right?

No in the sense that some code could have been relying on the `sysWindowResizeEvent` to know when the profile was shown/hidden when multi-playing but multi-view mode was not switched on rather then using that signal to redraw because the screen dimensions were changed. This was what I commented on in #5887 when I "fixed" the former event so that it and the NAWS signalling were not being done when a profile was shown/hidden but only when there *was* a resize.

 This PR comes about because I foresaw that there could/would be a need to know when a profile gained/lost focus when multiplying - however before I got to the point it is at now Leris/Kebap pointed out that the Mudlet application as a whole was not properly keeping track of which profile was "in focus" (because the title bar was not being updated to reflect the active one) and that solving one should fix both things.

>#6449 seems unaffected / not solved with this.

After this PR clicking on the `TConsole`/(pair of `TTextEdit`s in it) of a UserWindow (i.e. the floating/dockable windows) the focus will stay in the `TCommandLine` of that UserWindow if it has been enabled; it will also stay there as well when clicking within that same `TCommandLine`. However I have found that the same is NOT true of MiniConsoles (those that are embedded within the `TMainConsole`) as I realised I have missed one chunk of code, that being a `mouseReleaseEvent(QMouseEvent*)` for the `TCommandLine` class.

Furthermore, the focus largely stays on a secondary (additional) command line added to the main console, though that is also a bit off because of the aforesaid missing release event handler. OTOH Clicking on the floating/dockable mapper and both the buttons/menus AND any empty space in a floating/dockable toolbar returns the focus to the main command line and sets the active profile to the correct one - before this PR clicking on empty space in such a toolbar didn't have any effect.

vadi2 commented:
> some code could have been relying on the sysWindowResizeEvent to know when the profile was shown/hidden

Is there such code though? The original code linked in the issue wasn't doing of that, it was just broken due to wrong data being reported to it 🤔

atari2600tim commented:
For #6450, clicking inside of map while on different profile, this does seem to fully switch over to the other profile properly.

SlySven committed:
Fix: add mouseReleaseEvent handler for commandline

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Waiting on: #6451 ...

SlySven commented:
Oops,  it seems we don't have a "This PR works but is waiting for another PR to fix the `development` codebase so that the CI builds work"...!

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

Needed to pull in fix for Linux CI builds from development codebase.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

vadi2 commented:
I'm still a bit unclear on the need for `sysProfileFocusChangeEvent`. What code is expected to make use of it and at what times, and why doesn't the existing `sysWindowResizeEvent` one cover it?

> No in the sense that some code could have been relying on the sysWindowResizeEvent to know when the profile was shown/hidden when multi-playing but multi-view mode was not switched on rather then using that signal to redraw because the screen dimensions were changed. 

Where can I find this code?

There are also semi-related fixes to focus tracking of userwindows and command lines - that would be better served and easier to merge if it was a separate PR, I reckon.

SlySven commented:
> I'm still a bit unclear on the need for `sysProfileFocusChangeEvent`. What code is expected to make use of it and at what times, and why doesn't the existing `sysWindowResizeEvent` one cover it?

It is foreseeable - at least I can see it - that when multi-playing a profile might need to know whether it has the player's attention/focus - especially if they are NOT using multi-view mode {imagine that the profile is not the one being shown and a monster or enemy attacks the character in that profile}. As such this new event provides something (a signal that the profile has gained OR lost focus) that a script/package can use for that. Previously they might have been able to use the `sysWindowResizeEvent` - but since #5887 that ***no longer fires*** when the user clicks between tabs/profiles so cannot be used to achieve it.

Originally I thought that the elimination of the bogus NAWS messages was the cause of #5960 - but it turns out it wasn't. At the same time though various parties, @Kebap included, noted that clicking on various UI elements of a profile was not causing the focus to be switched back to that profile if more than one was active; as such this PR morphed into something that could fix that as well - because the areas of the codebase involved are interrelated. As such this PR should clean things up because the process of switching between profiles has been streamlined IMHO.

SlySven commented on src/TConsole.h:
```diff
@@ -192,27 +192,24 @@ class TConsole : public QWidget
-
     ConsoleType getType() const { return mType; }
     virtual void setProfileName(const QString&);
-    // In the next pair of functions the first element in the return is an
```
TBH I spotted this comment error in passing and fixed it - I didn't feel up to doing a whole PR just for this.

SlySven commented on src/TConsole.h:
```diff
@@ -192,27 +192,24 @@ class TConsole : public QWidget
     QPointer<Host> mpHost;
     // Only assigned a value for user windows:
     QPointer<TDockWidget> mpDockWidget;
-    // Only on a MainConsole type instance:
```
It was true when I wrote it but it not longer is!

SlySven commented on src/mudlet.cpp:
```diff
@@ -4593,26 +4526,107 @@ void mudlet::setNetworkRequestDefaults(const QUrl& url, QNetworkRequest& request
+    mpTabBar->setStyleSheet(mpCurrentActiveHost->mProfileStyleSheet);
+    menuBar()->setStyleSheet(mpCurrentActiveHost->mProfileStyleSheet);
+
+    // Tell the old profile that it is losing focus via a Mudlet event:
```
Oops! A bit of a copy-pasta error:
```suggestion
    // Tell the new profile that it is gaining focus via a Mudlet event:
```

SlySven committed:
Update src/mudlet.cpp

vadi2 commented on src/mudlet.cpp:
```diff
@@ -1579,13 +1507,16 @@ void mudlet::addConsoleForNewHost(Host* pH)
 
     //update the main window title when we spawn a new tab
-    setWindowTitle(pH->getName() + " - " + version);
+    setWindowTitle(tr("%1 - %2",
```
Nothing for translators to translate here, isn't it? Should just be `qsl()`.

SlySven replied:
The spaces and the hyphen are probably not appropriate for some Far Eastern locales...!

It is also possible that the ordering might not be the same everywhere, i.e. the version number might be first in some places. :man_shrugging:

vadi2 replied:
I'm not certain about that... lets leave it be for now until folks ask and we know it's needed for sure.

SlySven replied:
Okay - I guess - for NOW.

:earth_africa:  **I18n**/**L10n** I say NOW because we currently force the user to call their profiles something that only uses a subset of ASCII characters (so that it is safe to use them as a directory name) - but long-term we ought to try and find a way around that so that a profile can be named in an end-user locale appropriate manner.

vadi2 commented:
Thanks, the need for the event makes sense! I see it's making use of 0's and 1's - could you describe how that works so @Mudlet/lua-interface can review it appropriately?

SlySven commented:
> Thanks, the need for the event makes sense! I see it's making use of 0's and 1's - could you describe how that works so @Mudlet/lua-interface can review it appropriately?

No need, the `0`/`1` are how we pass `false`/`true` boolean values (for "Lost"/"Gained" focus respectively) in the Mudlet `TEvent` code - by the time the users/@Mudlet/lua-interface get to see them they will be the latter Lua values...

SlySven committed:
Merge branch 'development' into Add_sysProfileFocusChange_event

SlySven committed:
Fix: correct a couple of merge errors

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:books: Provisional documentation in Area 51: https://wiki.mudlet.org/w/Area_51#sysProfileFocusChangeEvent.2C_PR6267_open.29 .

vadi2 commented on src/mudlet.cpp:
```diff
@@ -4483,26 +4400,107 @@ void mudlet::setNetworkRequestDefaults(const QUrl& url, QNetworkRequest& request
+    mpCurrentActiveHost->mpConsole->repaint();
+    mpCurrentActiveHost->mpConsole->refresh();
+    mpCurrentActiveHost->mpConsole->mpCommandLine->repaint();
+    // CHECKME: Although we no longer force the focus to the profiles main
```
Who can check this?

SlySven replied:
Well, I vaguely recall observing that the main command line was expanding/contracting (vertically) in a manner that did not always match up with what it contained - so presumably that would be affected by such repaint operations. If that is indeed the case then it could also apply to secondary / alternative command lines, so I'd hope that anyone who might be investigating that would note this...

vadi2 replied:
A comment on the relevant issue would be much easier to find in that case!

SlySven replied:
I haven't found an issue for it (yet) though - and I am not sure if what I saw constituted an actual issue. If there IS one, then steps to address it might need to include all command lines.

vadi2 replied:
Alright - if it was an actual problem, mentioning this in the real issue would make it easy to find when someone was going to tackle it.

But like this, where we're not even certain if we have a real problem, I don't see how it'll help anyone much at all.

SlySven replied:
Okay - I've shortened it to a one-liner comment that I think conveys the fundamentals... :stuck_out_tongue:

vadi2 commented:
Otherwise looks solid to me.

SlySven commented on src/mudlet.cpp:
```diff
@@ -1460,22 +1380,17 @@ void mudlet::addConsoleForNewHost(Host* pH)
-    // "version" to "scmVersion" for the second argument when this chunk of code
-    // gets moved to mudlet::activateProfile() - and remove this comment!
     // update the window title for the currently selected profile
     setWindowTitle(tr("%1 - %2",
                       // Intentional comment to separate arguments
                       "Title for the main window when a profile is loaded or active, %1 is the name "
                       "of the profile and %2 is the Mudlet version string.")
                            .arg(pH->getName(), scmVersion));
```
```suggestion
    // Potential to be translated in the future if the need arises, with the following disambiguation:
    // "Title for the main window when a profile is loaded or active, %1 is the name "
    // "of the profile and %2 is the Mudlet version string."
    setWindowTitle(qsl("%1 - %2"))
                           .arg(pH->getName(), scmVersion));
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -4483,26 +4400,107 @@ void mudlet::setNetworkRequestDefaults(const QUrl& url, QNetworkRequest& request
+    QApplication::sendEvent(mpCurrentActiveHost->mpConsole, &event);
+
+    // update the main application window title for the currently selected profile
+    setWindowTitle(tr("%1 - %2",
+                      // Intentional comment to separate arguments
+                      "Title for the main window when a profile is loaded or active, %1 is the name "
+                      "of the profile and %2 is the Mudlet version string.")
+                           .arg(mpCurrentActiveHost->getName(), scmVersion));
```
```suggestion
    // Potential to be translated in the future if the need arises, with the following disambiguation:
    // "Title for the main window when a profile is loaded or active, %1 is the name "
    // "of the profile and %2 is the Mudlet version string."
    setWindowTitle(qsl("%1 - %2"))
                           .arg(mpCurrentActiveHost->getName(), scmVersion));
```

SlySven reviewed (commented):
Taking out the translation of the main window title - but leaving it in a manner that can be reintroduced later...

SlySven committed:
Revise: remove main window title from texts to be translated

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/mudlet.cpp:
```diff
@@ -1460,22 +1380,17 @@ void mudlet::addConsoleForNewHost(Host* pH)
+    // Potential to be translated in the future if the need arises, with the following disambiguation:
+    // "Title for the main window when a profile is loaded or active, %1 is the name "
+    // "of the profile and %2 is the Mudlet version string."
+    setWindowTitle(qsl("%1 - %2"))
```
```suggestion
    setWindowTitle(qsl("%1 - %2")
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -4483,26 +4400,107 @@ void mudlet::setNetworkRequestDefaults(const QUrl& url, QNetworkRequest& request
+    // Potential to be translated in the future if the need arises, with the following disambiguation:
+    // "Title for the main window when a profile is loaded or active, %1 is the name "
+    // "of the profile and %2 is the Mudlet version string."
+    setWindowTitle(qsl("%1 - %2"))
```
```suggestion
    setWindowTitle(qsl("%1 - %2")
```

SlySven committed:
Fix: typos in prior commit

SlySven commented on src/mudlet.cpp:
```diff
@@ -4483,26 +4400,107 @@ void mudlet::setNetworkRequestDefaults(const QUrl& url, QNetworkRequest& request
+    mpCurrentActiveHost->mpConsole->repaint();
+    mpCurrentActiveHost->mpConsole->refresh();
+    mpCurrentActiveHost->mpConsole->mpCommandLine->repaint();
+    // CHECKME: Although we no longer force the focus to the profiles main
+    // console's command line here (because it isn't always appropriate) we are
+    // redrawing the main main console AND it's primary command line. That
+    // being the case should we perhaps also be repainting any "mini consoles"
+    // and/or any additional command lines, that have either been added to the
+    // main console or enabled on any mini console or "user window"?
```
```suggestion
    // CHECKME: Why are we only repainting the main command line, what about the others?
```

SlySven committed:
Revise: shorten an overlong comment

vadi2 reviewed (approved):
Thanks for this 🙏
